File: 51.n-queens.py
#
# @lc app=leetcode id=51 lang=python3
#
# [51] N-Queens
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []

        def dfs(r, path, cols, pdiag, ndiag):
            for c in range(n):
                if (cols & (1 << c)) or (pdiag & (1 << (r + c))) or (ndiag & (1 << (r - c + n))):
                    continue

                if r == n - 1:
                    path.append(c)
                    res.append("." * c + "Q" + "." * (n - c - 1) for c in path)
                    return

                dfs(
                    r + 1,
                    path + [c],
                    cols | 1 << c,
                    pdiag | 1 << (r + c),
                    ndiag | 1 << (r - c + n),
                )

        dfs(0, [], 0, 0, 0)
        return res


# @lc code=end

File: 51.n-queens.rs
/*
 * @lc app=leetcode id=51 lang=rust
 *
 * [51] N-Queens
 */

// @lc code=start
impl Solution {
    pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
        let mut res = vec![];

        fn dfs(
            r: usize,
            path: &mut Vec<String>,
            cols: u16,
            pdiag: u16,
            ndiag: u16,
            res: &mut Vec<Vec<String>>,
            n: usize,
        ) {
            for c in 0..n {
                if (cols & (1 << c)) != 0
                    || (pdiag & (1 << (r + c))) != 0
                    || (ndiag & (1 << (r - c + n))) != 0
                {
                    continue;
                }
                if r == n - 1 {
                    path.push(format!("{}Q{}", ".".repeat(c), ".".repeat(n - c - 1),));
                    res.push(path.clone());
                    path.pop();
                    return;
                }

                path.push(format!("{}Q{}", ".".repeat(c), ".".repeat(n - c - 1),));
                dfs(
                    r + 1,
                    path,
                    cols | (1 << c),
                    pdiag | (1 << (r + c)),
                    ndiag | (1 << (r - c + n)),
                    res,
                    n,
                );
                path.pop();
            }
        }

        dfs(0, &mut vec![], 0, 0, 0, &mut res, n as usize);
        res
    }
}
// @lc code=end

